src/cpp/dds/domain/DomainParticipantFactory.cpp

```diff
@@ -0,0 +1,250 @@
+            {
+                pit->disable();
+            }
+        }
+        // Delete participants
+        for (auto it : participants_)
+        {
+            for (auto pit = it.second.begin(); pit != it.second.end(); ++pit)
+            {
+                delete *pit;
+            }
+        }
+        participants_.clear();
+    }
+
+    // Deletes DynamicTypes and TypeObject factories
+    fastrtps::types::DynamicTypeBuilderFactory::delete_instance();
+    fastrtps::types::DynamicDataFactory::delete_instance();
+    fastrtps::types::TypeObjectFactory::delete_instance();
+    XMLProfileManager::DeleteInstance();
+
+    fastrtps::eClock::my_sleep(100);
+    fastrtps::Log::KillThread();
+}
+
+DomainParticipantFactory* DomainParticipantFactory::get_instance()
+{
+    if (!g_instance_initialized)
+    {
+        std::lock_guard<std::mutex> lock(g_mtx);
+        if (g_instance == nullptr)
+        {
+            g_instance = new DomainParticipantFactory();
+            g_instance_initialized = true;
+        }
+    }
+    return g_instance;
+}
+
+bool DomainParticipantFactory::delete_instance()
+{
+    std::lock_guard<std::mutex> lock(g_mtx);
+    if (g_instance_initialized && g_instance != nullptr)
+    {
+        delete g_instance;
+        g_instance = nullptr;
+        g_instance_initialized = false;
+        return true;
+    }
+    return false;
+}
+
+bool DomainParticipantFactory::delete_participant(
+        DomainParticipant* part)
+{
+    if (part != nullptr)
+    {
+        std::lock_guard<std::mutex> guard(mtx_participants_);
+
+        auto it = participants_.find(part->get_domain_id());
+
+        if (it != participants_.end())
+        {
+            auto pit = std::find(it->second.begin(), it->second.end(), part->impl_);
+            if (pit != it->second.end())
+            {
+                it->second.erase(pit);
+                delete *pit;
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+DomainParticipant* DomainParticipantFactory::create_participant(
+        const std::string& participant_profile,
+        DomainParticipantListener* listen)
+{
+    if (false == default_xml_profiles_loaded)
+    {
+        XMLProfileManager::loadDefaultXMLFile();
+        default_xml_profiles_loaded = true;
+    }
+
+    ParticipantAttributes participant_att;
+    if ( XMLP_ret::XML_ERROR == XMLProfileManager::fillParticipantAttributes(participant_profile, participant_att))
+    {
+        logError(DOMAIN_PARTICIPANT_FACTORY, "Problem loading profile '" << participant_profile << "'");
+        return nullptr;
+    }
+
+    return create_participant(participant_att, listen);
+}
+
+DomainParticipant* DomainParticipantFactory::create_participant(
+        const ParticipantAttributes& att,
+        DomainParticipantListener* listen)
+{
+    uint8_t domain_id = static_cast<uint8_t>(att.rtps.builtin.domainId);
+    DomainParticipant* pubsubpar = new DomainParticipant();
```

Comment:
Maybe `participant` or `domain_participant` would be a better name for this variable. 
The same with impl one